### PR TITLE
Auditorium seated users refactor

### PR DESCRIPTION
**Pull request by dmitrymarchuk**
This PR ensures that New Auditorium:

- does not use `useRecentVenueUsers` anymore
- stores its seated users inside `seatedUsers` subcollection of `sections` documents
- takes user count inside each section and users sample for avatars from `seatedUsersCount` and `seatedUsersSample` inside `sections` documents

`seatedUsersCount` and `seatedUsersSample` are updated with scheduled cloud function.

### CODE REVIEW

File: src/hooks/useTraceUpdate.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { useEffect, useRef } from "react";
+
+interface ObjectDict {
+  [index: string]: unknown;
+}
+
+export const useTraceUpdate = (props: Record<string, unknown>) => {
```
Comment (dmitrymarchuk) [inline src/hooks/useTraceUpdate.ts:7]: it's just a useful hook for debugging purposes.

File: functions/scheduled.js [hunk 2]
```diff
@@ -3,10 +3,74 @@ const admin = require("firebase-admin");
...
 
-const { chunk, sampleSize } = require("lodash");
+const { chunk, sampleSize, flatMap } = require("lodash");
 const hoursToMilliseconds = require("date-fns/hoursToMilliseconds");
 
 const DEFAULT_RECENT_USERS_IN_VENUE_CHUNK_SIZE = 6;
+const SECTION_PREVIEW_USER_DISPLAY_COUNT = 14;
+
+exports.updateSeatedUsersCountInAuditorium = functions.pubsub
+  .schedule("every 5 seconds")
```
Comment (mike-lvov) [inline functions/scheduled.js:13]: Reduce it to 5 minutes to reduce the costs

Comment (mike-lvov) [reply functions/scheduled.js:13]: Btw, in this exact scheduled function when you fetch seated users you can delete the old ones

File: src/api/venue.ts [hunk 3]
```diff
@@ -77,3 +79,65 @@ export const updateIframeUrl = async (iframeUrl: string, venueId?: string) => {
...
+    .doc(venueId)
+    .collection("sections")
+    .doc(sectionId)
+    .collection("seatedUsers")
+    .doc(userId);
+
+export const unsetAuditoriumSectionSeat = async (
+  userId: string,
+  venueId: string,
+  sectionId: string
```
Comment (mike-lvov) [inline src/api/venue.ts:100]: Given that all of those are string, use an object here

File: src/api/venue.ts [hunk 4]
```diff
@@ -77,3 +79,65 @@ export const updateIframeUrl = async (iframeUrl: string, venueId?: string) => {
     .functions()
     .httpsCallable("venue-adminUpdateIframeUrl")({ venueId, iframeUrl });
 };
+
+const getUserInSectionRef = (
+  userId: string,
+  venueId: string,
+  sectionId: string
```
Comment (mike-lvov) [inline src/api/venue.ts:86]: use an object here

File: src/api/venue.ts [hunk 5]
```diff
@@ -77,3 +79,65 @@ export const updateIframeUrl = async (iframeUrl: string, venueId?: string) => {
...
+export const setAuditoriumSectionSeat = async (
+  user: WithId<DisplayUser>,
+  venueId: string,
+  sectionId: string,
+  row: number,
+  column: number
+) => {
+  return getUserInSectionRef(user.id, venueId, sectionId)
+    .set({
+      ...pickDisplayUserFromUser(user),
```
Comment (mike-lvov) [inline src/api/venue.ts:127]: why do you want to spread its data? Why not just store it as `user`

File: src/hooks/auditorium/useAllAuditoriumSections.ts [hunk 6]
```diff
@@ -62,16 +58,10 @@ export const useAllAuditoriumSections = (venue: WithId<AuditoriumVenue>) => {
...
       const sectionCapacity = getSectionCapacity(venue, section);
-      const seatedUsers = getAuditoriumSeatedUsers({
-        venueId,
-        // @debt should be replaced with a subcollection
-        auditoriumUsers: recentVenueUsers,
-        sectionId: section.id,
-      });
 
-      return seatedUsers.length < sectionCapacity;
+      return (section?.seatedUsersCount ?? 0) < sectionCapacity;
```
Comment (mike-lvov) [inline src/hooks/auditorium/useAllAuditoriumSections.ts:62]: move seatedUsersCount to a var

File: src/hooks/useAuditoriumSeatedUsers.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  emptyArray,
+} from "utils/selectors";
+
+import { useFirestoreConnect } from "hooks/useFirestoreConnect";
+import { useSelector } from "hooks/useSelector";
+
+export const findAuditoriumSeatedUsers = (
+  recentVenueUsers: readonly WithId<User>[],
+  venueId: string | undefined,
+  sectionId: string | undefined
```
Comment (mike-lvov) [inline src/hooks/useAuditoriumSeatedUsers.ts:15]: props as an object

File: src/hooks/useAuditoriumSeatedUsers.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+) => {
+  return sectionId && venueId
+    ? recentVenueUsers.filter(
+        (user) => user.data?.[venueId]?.sectionId === sectionId
+      )
```
Comment (mike-lvov) [inline src/hooks/useAuditoriumSeatedUsers.ts:20]: do we need to keep this logic? As we now store them in a subcollection

File: src/hooks/useAuditoriumSeatedUsers.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        ],
+        storeAs: "currentAuditoriumSectionSeatedUsers",
+      },
+    ];
+  });
+};
+
+export const useAuditoriumSeatedUsers = (
+  venueId: string | undefined,
+  sectionId: string | undefined
```
Comment (mike-lvov) [inline src/hooks/useAuditoriumSeatedUsers.ts:49]: object

File: src/hooks/useAuditoriumSeatedUsers.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+) => {
+  useConnectAuditoriumSectionSeatedUsers(venueId, sectionId);
+
+  return useSelector(currentAuditoriumSectionSeatedUsersSelector);
```
Comment (mike-lvov) [inline src/hooks/useAuditoriumSeatedUsers.ts:53]: return the isLoaded + ?? ALWAYS_EMPTY_ARRAY

File: src/api/venue.ts (hunk 5 above)
Comment (dmitrymarchuk) [reply src/api/venue.ts:127]: For `WithId` stuff in selectors to work properly
I've tried to make a separate field `user` - it complicates things.

File: functions/scheduled.js [hunk 11]
```diff
@@ -3,10 +3,74 @@ const admin = require("firebase-admin");
...
+
+    const firestore = admin.firestore();
+    const auditoriums = await firestore
+      .collection("venues")
+      .where("template", "==", "auditorium")
+      .get();
+
+    const sectionsArray = await Promise.all(
+      auditoriums.docs.map(({ ref }) => ref.collection("sections").get())
+    );
```
[the remark this answers is not in the document]
Comment (dmitrymarchuk) [reply functions/scheduled.js]: Rewritten in https://github.com/sparkletown/sparkle/pull/2318/commits/31f0c55b5db370cbcbe5998ea7ab45d54b0916fb and https://github.com/sparkletown/sparkle/pull/2318/commits/787c2e649077534aa9bc2668da5f25350191253b
The code haven't become much simpler but surely the amount of `await`ing and requests dropped dramatically.

File: functions/scheduled.js (hunk 2 above)
Comment (dmitrymarchuk) [reply functions/scheduled.js]: fixed in https://github.com/sparkletown/sparkle/pull/2318/commits/31f0c55b5db370cbcbe5998ea7ab45d54b0916fb

Comment (dmitrymarchuk) [reply functions/scheduled.js]: we will cope with "dangling" users in a separate PR

File: src/api/venue.ts [hunk 12]
```diff
@@ -77,3 +79,65 @@ export const updateIframeUrl = async (iframeUrl: string, venueId?: string) => {
...
+      throw err;
+    });
+};
+
+export const setAuditoriumSectionSeat = async (
+  user: WithId<DisplayUser>,
+  venueId: string,
+  sectionId: string,
+  row: number,
+  column: number
```
[the remark this answers is not in the document]
Comment (dmitrymarchuk) [reply src/api/venue.ts]: fixed in https://github.com/sparkletown/sparkle/pull/2318/commits/f0c98b720e24314deebbf7593d97ad53147ba0d3

File: src/api/venue.ts (hunk 3 above)
Comment (dmitrymarchuk) [reply src/api/venue.ts]: fixed in https://github.com/sparkletown/sparkle/pull/2318/commits/f0c98b720e24314deebbf7593d97ad53147ba0d3

File: src/api/venue.ts (hunk 4 above)
Comment (dmitrymarchuk) [reply src/api/venue.ts]: fixed in https://github.com/sparkletown/sparkle/pull/2318/commits/f0c98b720e24314deebbf7593d97ad53147ba0d3

File: src/hooks/auditorium/useAllAuditoriumSections.ts (hunk 6 above)
Comment (dmitrymarchuk) [reply src/hooks/auditorium/useAllAuditoriumSections.ts]: fixed in https://github.com/sparkletown/sparkle/pull/2318/commits/10fe3f04176fbe949d0025a18c5ea6761b3b44d8

File: src/hooks/useAuditoriumSeatedUsers.ts (hunk 7 above)
Comment (dmitrymarchuk) [reply src/hooks/useAuditoriumSeatedUsers.ts]: fixed in https://github.com/sparkletown/sparkle/pull/2318/commits/10fe3f04176fbe949d0025a18c5ea6761b3b44d8

File: src/hooks/useAuditoriumSeatedUsers.ts (hunk 8 above)
Comment (dmitrymarchuk) [reply src/hooks/useAuditoriumSeatedUsers.ts]: yep, removed the entire method in https://github.com/sparkletown/sparkle/pull/2318/commits/10fe3f04176fbe949d0025a18c5ea6761b3b44d8

File: src/hooks/useAuditoriumSeatedUsers.ts (hunk 9 above)
Comment (dmitrymarchuk) [reply src/hooks/useAuditoriumSeatedUsers.ts]: fixed in https://github.com/sparkletown/sparkle/pull/2318/commits/10fe3f04176fbe949d0025a18c5ea6761b3b44d8

File: src/hooks/useAuditoriumSeatedUsers.ts (hunk 10 above)
Comment (dmitrymarchuk) [reply src/hooks/useAuditoriumSeatedUsers.ts]: fixed in https://github.com/sparkletown/sparkle/pull/2318/commits/bdb561ea34984e9f2dbe47a185fa73c904225250

File: src/hooks/useGetUserByPosition.ts [hunk 13]
```diff
@@ -12,43 +11,60 @@ export interface UseGetUserByPositionProps {
...
+
+        return {
+          ...user,
+          position: {
+            row,
+            column,
+          },
+        };
+      })
+    : [];
```
Comment (mike-lvov) [inline src/hooks/useGetUserByPosition.ts:37]: Does he still have this data?

File: src/types/User.ts [hunk 14]
```diff
@@ -2,12 +2,15 @@
...
 export interface Experience {
   // @debt refactor bartender to be potentially undefined. Or can we remove it entirely?
   bartender: User;
   table?: string | null;
-  row?: number | null;
-  column?: number | null;
+  row?: number;
+  column?: number;
   sectionId?: string;
 }
```
Comment (mike-lvov) [inline src/types/User.ts:16]: Hopefully we'll get rid of this in the next day or two

**Review by mike-lvov — APPROVED** [review]
LGTM

File: src/hooks/useGetUserByPosition.ts (hunk 13 above)
Comment (dmitrymarchuk) [reply src/hooks/useGetUserByPosition.ts:37]: I've renamed the old hook to `useGetUserByPositionOld` as it's still being used inside `Map.tsx`
But probably will get rid of that too when we finish with other venues.

**Pull request merged**